### ISSUE TITLE
Proposal for dealing with shared/derived data

### ISSUE THREAD

**Issue by abaco** [staltz/cycle-onionify#11]
**--- UPDATE: SEE COMMENTS BELOW FOR A BETTER PROPOSAL ---**

I've created a [branch](https://github.com/abaco/cycle-onionify/tree/optics) to experiment with lensed (or, more precisely, *optic'd*) state. It uses partial.lenses. The idea is to retain the scope-based isolation of the state object, while allowing to associate an optic (lens, traversal, or isomorphism) to a scope name, so that children components can see the state through said optics. 

For example one can have the following state:
```js
const reducer$ = xs.of({kelvin: 283});
```
and compute derived values through:
```js
const optics$ = xs.of({
  // this component sees the state as it is
  self: L.identity,
  children: {
    kelvin: {
      // a child isolated with 'kelvin' sees 283 as state
      self: L.prop('kelvin')
    },
    celsius: {
      // a child isolated with 'celsius' sees 9.85 as state
      self: L.compose(
        L.prop('kelvin'),
        L.iso(a => a - 273.15, b => b + 273.15)
      )
    },
    fahrenheit: {
      // a child isolated with 'fahrenheit' sees 49.73 as state
      self: L.compose(
        L.prop('kelvin'),
        L.iso(a => a * 9/5 - 459.67, b => (b + 459.67) * 5/9)}
      )
    }
  }
});
```
The component must then return `{..., onion: {optics$, reducer$}}` instead of the usual `{..., onion: reducer$}` (which however is still supported). The children components access the state through the optics (in this case isomorphisms) in a transparent way: they don't need to know about partial.lenses and they can update the value.

I've added two examples, "mixed-optics" and "mixed-optics-nested", which use slightly different approaches. The differences with the original "mixed" example are minimal.

This code is just a proof of concept. It's probably buggy and I haven't thought of performance at all. I'd just like to know what the community thinks of such an approach.

I think the pros of this solution are:
- Better separation of concerns and reusability of components. In the new "mixed" examples, each list item sees the counter value as if it owned it.
- Backwards compatibility.
- Power. This is the first time I use partial.lenses, and so far I'm really impressed. It seems extremely powerful.

**Comment by staltz** [comment]
:+1: to the overall direction and problems tackled (reusability and derivation, "each list item sees the counter value as if it owned it").

On the specifics (e.g. component must then return `{..., onion: {optics$, reducer$}}`), you'll need to give me some time to think about it myself and experiment with it too. I distribute my time with many other standing issues, so this may take some while.

**Comment by abaco** [comment]
Cool! Take your time, in the meanwhile I'll write some documentation and maybe another representative example.

**Comment by abaco** [comment]
I created another [branch](https://github.com/abaco/cycle-onionify/tree/optics-cleaner) with a much simpler and cleaner solution, both in the API and the internals. Please forget about the other branch.

Now one can set a `scopeToOptic` property of the `onion` source, specifying a function from isolation scope to optic. The above example would thus become:
```js
sourcer.onion.scopeToOptic = scope => {
  switch(scope) {
  // this branch is redundant, can be omitted
  case 'kelvin':
    return L.prop('kelvin');
  // this one selects 'kelvin' and transforms it to/from celsius
  case 'celsius':
    return L.compose(
      L.prop('kelvin'),
      L.iso(a => a - 273.15, b => b + 273.15)
    );
  // this one selects 'kelvin' and transforms it to/from fahrenheit
  case 'fahrenheit':
    return L.compose(
      L.prop('kelvin'),
      L.iso(a => a * 9/5 - 459.67, b => (b + 459.67) * 5/9)}
    );
  }
}
```
This callback must be defined before creating the isolated children components, of course.

I chose a function instead of a hash because it allows to do complex things, especially when dealing with lists. The `mixed-optics` example is an instance of such an application.

**Comment by pixeleet** [comment]
@abaco I have tried a different approach and I simply use lenses based on the path the user is looking to access in the state:

Here's an excerpt of the passing test of this approach:
(osmosify is an onionify clone with ramda and most, request invite to private repo if you're curious of the nasty things I did :))
```
test('Should be able to isolate state reducers based on a reducer object', t => {
  let asserts = 0
  function Main(sources) {
    const expected = [
      { foo: 'bar' },
      { foo: 'bar', bar: { baz: 'foobar' } },
      { foo: 'bar', bar: { baz: 'cool' } }
    ]
    sources.state.state$.subscribe({
      next(state) {
        t.deepEqual(state, expected.shift())
        asserts += 1
        if (expected.length === 0) {
          t.is(asserts, 3)
          t.pass()
        }
      },
      error(e) { t.fail(e) },
      complete() {}
    })
    return {
      state: {
        foo: O.of((prevState) => 'bar'),
        bar: O.of((prevState) => ({ ...prevState, baz: 'foobar' })),
        'bar.baz': O.of(prevState => 'cool')
      }
    }
  }
  const wrapped = osmosify(Main)
  isolate(wrapped)({})
})
```

**Comment by abaco** [comment]
@pixeleet "osmosify", cool name :-)

From your test it's not clear what you can do with osmosify (or at least it's not clear to me).

Are your `state` source and sink affected by `isolate`?

Can you specify other optics besides lenses on some subfield of the state object, like your `'bar.baz'`? In [this example](https://github.com/abaco/cycle-onionify/blob/optics-cleaner/examples/unit-conversion/src/TemperatureApp.ts#L55-L57) I convert from Kelvin to Celsius and Fahrenheit and vice versa. Can osmosify do that? I'm pretty sure stanga can, because there you can explicity lens a source before passing it to a child component.

How would the ["mixed" example](https://github.com/staltz/cycle-onionify/tree/master/examples/mixed) look like in osmosify? In [my version](https://github.com/abaco/cycle-onionify/blob/optics-cleaner/examples/mixed-optics/src/Item.ts#L47) the root state is:
```js
{
   list: [
      {content: 'A'},
      {content: 'B'}
   ],
   counter: {count: 3}
}
```
but each list item sees a state like
```js
{
   content: 'A',
   count: 3
}
```
and can modify `count` as if it owned it. Can osmosify do that? My undestanding is that in stanga it's not possible (but maybe @milankinen can comment on this).

**Comment by davidskuza** [comment]
@abaco The only one problem I see with your approach (2nd one) is that now you have to know every case of isolation at the root.

**Comment by abaco** [comment]
@davidskuza In fact that's not the case: at each level you can associate a custom lens to a specific isolation scope. Each level only need to know its children's isolations scopes, as usual.

In the "mixed" example the [TodoApp](https://github.com/abaco/cycle-onionify/blob/optics-cleaner/examples/mixed-optics/src/TodoApp.ts#L80-L85) component provides a lens that passes the whole state to its child isolated with `list`.

In turn the [List](https://github.com/abaco/cycle-onionify/blob/optics-cleaner/examples/mixed-optics/src/List.ts#L28-L45) component maps numerical isolation scopes (the indices of its items) to lenses that merge the global counter and the todo content.

**Comment by davidskuza** [comment]
Ah right, everytime we select new StateSource is created. I like your idea of lenses but I'm curious if we could make it non external library dependent, for example two functions for transformation and we could use whatever we want to (or just plain JS). Not sure if this is the right idea though.

**Comment by abaco** [comment]
I think we need partial.lenses (or something with similar features, if it exists) to handle a sufficiently wide range of combination/transformation of data. For instance I wouldn't have a clue how else to *express* something like [this](https://github.com/abaco/cycle-onionify/blob/optics-cleaner/examples/mixed-optics/src/List.ts#L31-L43), let alone make it work.

That said, we can maybe have partial.lenses as an optional dependency. In the current implementation lenses are internally created even for the usual cases, when `scopeToOptic` is not defined, but that could be changed. Then onionify would need to handle lenses only when they are provided by the user. I don't know how optional dependency work in npm, but in theory that seems doable.

**Comment by abaco** [comment]
Sorry, now maybe I understand what you meant. You want to make a sort of adapter for any lens library, right? That sound neat in theory, but in practice i find it overkill because:

- it would add a little complexity on the user's side
- the whole partial.lenses is 10 KB when minified
- with tree-shaking any KB difference between partial.lenses and ad hoc JS becomes negligible
- simple cases are very easy to handle with partial.lenses, you don't need to learn much
- I don't know of any library that can replace partial.lenses for the complex cases

**Comment by davidskuza** [comment]
For me lenses are good if no perfect but not for everyone that might be a case (on entry level at least, I'm thinking about people who are responsible for picking software stack for new projects. They see this first time, they don't understand in 3 seconds, they tell this is stupid...). Also not everybody might like the idea that you need that.

I'm kinda thinking about adapter but that could be only two functions instead of one lens. For example:

```
sources.onion.scopeTo = scope => {
  const index = parseInt(scope)

  if (!isNan(index)) {
    return {
      get(state) {
        return {
          content: state.list[index].content,
          count: state.counter.count
        }
      },
      set(state, newState) {
        state.list[index].content = newState.content
        state.counter.count = newState.count

        return state
      }
    }
  }
}
```

and 

```
sources.onion.scopeTo = scope => {
  const index = parseInt(scope)

  if (!isNan(index)) {
    const lenses = L.pick({
      content: L.compose(
        L.prop('list'),     // select 'list' property
        L.required([]),     // prevent removal of 'list' when it has no element
        L.index(index),     // select ith element of array
        L.prop('content')   // select 'content' property
      ),
      count: L.compose(
        L.prop('counter'),  // select 'counter' property
        L.prop('count'),    // select 'count' property
        L.choose(L.define)  // allow updates of counter.count, but prevent removal
      )
    })

    return {
      get: state => L.get(lenses),
      set: state => L.set(lenses)
    }
  }
}
```

Not sure if this is better though, just writing my thoughts.

**Comment by abaco** [comment]
Actually, I like your idea very much! The overhead compared to the partial.lenses dependent version would be really minimal.

At this point I tend to prefer your approach. I don't have much time these days, but I'll try make a working version in a separate branch.

**Comment by abaco** [comment]
@davidskuza I did as you said! There's a [new branch](https://github.com/abaco/cycle-onionify/tree/explicit-isolation) that implements your suggestions.

There are four examples that use the new feature:
- [`mixed-accessors`](https://github.com/abaco/cycle-onionify/blob/explicit-isolation/examples/mixed-accessors/src/List.ts#L27-L52) is a version of the `mixed` example that uses getters and setters to share data among sibling components
- [`mixed-optics`](https://github.com/abaco/cycle-onionify/blob/explicit-isolation/examples/mixed-optics/src/List.ts#L29-L46) same but with partial.lenses
- [`unit-conversion`](https://github.com/abaco/cycle-onionify/blob/explicit-isolation/examples/unit-conversion/src/TemperatureApp.ts#L25-L50) uses getters and setters to create derived data
- [`unit-conversion-optics`](https://github.com/abaco/cycle-onionify/blob/explicit-isolation/examples/unit-conversion-optics/src/TemperatureApp.ts#L27-L46) same but with partial.lenses

Im happy with this solution because it's not dependent on partial.lenses, but it's very easy to use with it.

**Comment by davidskuza** [comment]
You were faster than I thought, I thought I'll come with it first :smile:

This is great!
I'm happy with this solution also and probably will go with this in my projects. Didn't saw better ideas.

Now what @staltz thinks about that (or other users).

**Comment by staltz** [comment]
I like the purpose and the new capabilities that optics enables, but I'm still looking for an API. I'm not so sold on mutating `sources.onion.scopeToAccessors`. I think you guys could publish another library, or a fork of this library in the mean time. Usually I acknowledge the need, but I ruminate on the API design for a long while until I settle on something. That's how e.g. `isolate` came to existence. I'm still in the ruminating phase when it comes to advanced optics. :)

**Comment by abaco** [comment]
@staltz Take your time and ruminate as long as you need :) We'll probably publish a fork in the meanwhile.

Regarding your concern, mutating the `StateSource` is not necessary at all. We could do something like:
```js
const lensingOnionSource = sources.onion.overrideIsolation(scope => {get: ..., set: ...});
```
which would return a new source with the `scopeToAccessors` callback set. Then one would pass the new source to the child components. Does that look better to you? (In fact it does look better to me).

Besides the API and the mutation issue, I think we have no choice to implement lensing other than:

A. provide each child with a modified source (what stanga and culli do)
B. modify the source of the parent component, and pass the same source to all the children

Option B fits better with onionify's reliance on `isolate`.

**Comment by abaco** [comment]
Actually there's a third option:

C. provide `isolate` with the information to perform lensing

To that end I updated `isolate` (cyclejs/cyclejs#526) so that different drivers/wrappers can get each one a different scope, and scopes are not limited to strings. Here [yet another branch](https://github.com/abaco/cycle-onionify/tree/complex-lens-scope).

This is how you use it:
```js
const celsiusLens = {
  get: state => ({
    count: state.kelvin.count - 273.15
  }),
  set: (state, childState) => ({
    kelvin: {
      count: childState.count + 273.15
    }
  })
}

const celsiusSinks = isolate(Child, {onion: celsiusLens})(sources)
```

This option is not as weird as it might seems. We already pass numbers, instead of strings, as scopes, just to have them internally converted to strings by isolate and back to number by onionify. I think it's limiting to force scope to be strings: they are instructions on how to perform isolation. In most case one doesn't care at all of how isolation is done, and so lets `isolate` generate a random scope; but in some cases, like with onionify, you want more control, and there's no such thing as too much control.

**Comment by staltz** [comment]
> I think it's limiting to force scopes to be strings: they are instructions on how to perform isolation

I like this thinking. I like it a lot. Helps my rumination :)

**Comment by abaco** [comment]
I'm also ruminating, a lot... :)

When you first came out with onionify I liked the idea of piggybacking on `isolate` to select parts of the state tree. It *was* piggybacking, because `isolate` wasn't invented for such a purpose, but it didn't feel like a hack because it somehow fit well. Now I know why onionify is not a hack: because scopes are "isolation instructions", they've always been, also in the other drivers. Onionify was only the first to exploit this fact, although not to its fullest potential - for that we need to move beyond strings.

I'm not completely sure this is all correct, however, because I don't have the full picture. I don't exactly know how isolation works in the other drivers. But these are my thoughts, and if they're correct there might be consequences for the other drivers as well.

In onionify scopes are not only used to isolate components from each other, but also to *connect* them. This happens when a component knows the scope of one of its descendants. In such cases ancestor and descendant can communicate. Now my question is: would that be possible/useful also in other drivers? For instance, could that provide a solution to cyclejs/cyclejs#441?

Another line of thought concerns the manipulation of sources and sinks. Can that be useful in other drivers besides onionify? What if you could freely transform the DOM source and sink? A DOM source manipulation could for example throttle the events. Or maybe one could include an HTML component in a react-native app by transorming its DOM source and sink.

**Comment by staltz** [comment]
> Now my question is: would that be possible/useful also in other drivers? For instance, could that provide a solution to cyclejs/cyclejs#441?

*Possibly*, but I'm not sure how. That issue is about opting-out from parent-child isolation. How do we instruct the DOM source/sink of a child to either do "sibling isolation" or "parent-child isolation" or "both"? What kind of API we can use for that which naturally fits well and also is self-explanatory?

> Another line of thought concerns the manipulation of sources and sinks. Can that be useful in other drivers besides onionify?

Manipulation of sources and sinks is something pretty easy to do since they are inputs and outputs to a function. It can be done manually, or you can make a function that wraps a child component and does this manipulation. These functions are "component utilities" (a term @TylorS came up with, explained here: https://github.com/cyclejs/cyclejs/wiki/Structuring-Larger-Applications-(Guidelines) ). 

`isolate` is a component utility. It takes a component and returns a component. Internally, it manipulates sources and sinks. `onionify` is also another component utility. So I would say it has nothing to do with drivers (it may or may not add a new type of source/sink. `isolate` doesn't add for instance). It's simply a way of adding more logic to your program by wrapping some parts. I wish I had made many component utility libraries in order to set the example so people could see this useful pattern more often. For instance for transforming DOM sinks, this is a great example https://github.com/cyclejs-community/cyclejs-sortable/. It's not exactly a component utility, but it *could* be, if it had a different API.

**Comment by staltz** [comment]
By the way, I'm pretty much convinced we should do `isolate(Child, {onion: lensObj})` as well as https://github.com/cyclejs/cyclejs/issues/526. It's not a breaking change, the API is simple and intuitive, and it doesn't seem like it would add an obstacle if we want to consider how to solve https://github.com/cyclejs/cyclejs/issues/441. 👍 

Would be nice to get a third opinion before going forward, though.

**Comment by pixeleet** [comment]
One of the things that I ran into while using onion is that sometimes I really wanted the component to _just_ have the _root_ state instead of a random isolation scope like `cycle2` for which this seems to provide a much nicer api and thus enable non-explicitly scoped isolations to work well too :)

**Comment by abaco** [comment]
@staltz Glad that you approve the proposal! And thank you for the explanation on component utilities. I wasn't thinking of `onionify` as a component utility.

I understand that arbitrary manipulations of sources and sinks don't belong to `isolate`. But *some* kinds of manipulations might. I still don't have a single clear concept of isolation that covers the whole range of use cases from DOM to onionify... but that's my problem :) After all you're right, we don't have to solve cyclejs/cyclejs#441 here.

However, I've noticed that this change *does* break one of the examples. The problem is that numerical scopes aren't converted into strings any more, and so all drivers get numerical scopes while they expect strings. For some reason that caused issues with the first item in the `advanced` example, so I had to modify [this](https://github.com/abaco/cycle-onionify/commit/9ba7d5e981fe10910314c5b04b05af3f8dc9ea3b#diff-24135e224aeaec19e0deb2838941384eL25).

**Comment by abaco** [comment]
> I've noticed that this change does break one of the examples

That's just a consequence of how I modified `isolate`, but it can be avoided. I removed the stringification of scopes in all cases. My mistake.

If `isolate` stringifies the scope when called the usual way (e.g. `isolate(Component, 3)`) then nothing will break. Onionify will just have to handle both the following cases:
```js
isolate(Component, 3) // onionify gets '3' like all the other channels
isolate(Component, {onion: 3}) // onionify gets 3
```

**Comment by staltz** [comment]
Can you explain in more details why would `{onion: 3}` interpreted as `{onion: '3'}` break functionality?

**Comment by abaco** [comment]
Who would interpret it as `{onion: '3'}`? It can't be `isolate`, because it must pass through all the scopes given with the new syntax, whether they are `null`, numbers, lens objects or whatever else. If instead you mean that onionify interprets `{onion: 3}` as `{onion: '3'}`, then of course it can do it and it's no problem.

We need `isolate` to accept scopes of any type, but we also have to keep the stringification of scopes passed the usual way (e.g. `isolate(Component, 3)`), otherwise current application code that uses numerical scopes will break.

**Comment by staltz** [comment]
Right, understood.

**Comment by staltz** [comment]
Done! Released in v3.1.0
